File: 63. Unique Paths II/63. Unique Paths II.js
function uniquePathsWithObstacles(obstacleGrid) {
    const m = obstacleGrid.length;
    const n = obstacleGrid[0].length;
    
    // If the start or end cell is an obstacle, return 0
    if (obstacleGrid[0][0] === 1 || obstacleGrid[m-1][n-1] === 1) {
        return 0;
    }
    
    // Initialize the dp table
    const dp = Array.from({ length: m }, () => Array(n).fill(0));
    
    // Base case
    dp[0][0] = 1;
    
    // Fill the first row
    for (let j = 1; j < n; j++) {
        dp[0][j] = obstacleGrid[0][j] === 0 ? dp[0][j-1] : 0;
    }
    
    // Fill the first column
    for (let i = 1; i < m; i++) {
        dp[i][0] = obstacleGrid[i][0] === 0 ? dp[i-1][0] : 0;
    }
    
    // Fill the rest of the dp table
    for (let i = 1; i < m; i++) {
        for (let j = 1; j < n; j++) {
            if (obstacleGrid[i][j] === 0) {
                dp[i][j] = dp[i-1][j] + dp[i][j-1];
            } else {
                dp[i][j] = 0;
            }
        }
    }
    
    // The result is in the bottom-right corner of the dp table
    return dp[m-1][n-1];
}

File: 63. Unique Paths II/63. Unique Paths II.rb
# @param {Integer[][]} obstacle_grid
# @return {Integer}
def unique_paths_with_obstacles(obstacle_grid)
    m, n = obstacle_grid.length, obstacle_grid[0].length
    dp = Array.new(m) { Array.new(n, 0) }
    
    # Initialize first cell
    dp[0][0] = obstacle_grid[0][0] == 1 ? 0 : 1
    
    # Initialize first row
    (1...n).each do |j|
        dp[0][j] = (obstacle_grid[0][j] == 1 || dp[0][j-1] == 0) ? 0 : 1
    end
    
    # Initialize first column
    (1...m).each do |i|
        dp[i][0] = (obstacle_grid[i][0] == 1 || dp[i-1][0] == 0) ? 0 : 1
    end
    
    # Fill the dp table
    (1...m).each do |i|
        (1...n).each do |j|
            if obstacle_grid[i][j] == 1
                dp[i][j] = 0
            else
                dp[i][j] = dp[i-1][j] + dp[i][j-1]
            end
        end
    end
    
    dp[m-1][n-1]
end